client/worker.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2019 Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	jaegerpb "github.com/jaegertracing/jaeger/model"
+	ocTrancelator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+	"go.opencensus.io/trace"
+
+	sapmpb "github.com/signalfx/sapm-proto/gen"
+)
+
+const (
+	defaultRateLimitingBackoffSeconds = 8
+	headerAuthToken                   = "X-SF-Token"
+	headerRetryAfter                  = "Retry-After"
+	headerContentEncoding             = "Content-Encoding"
+	headerContentType                 = "Content-Type"
+	headerValueGZIP                   = "gzip"
+	headerValueXProtobuf              = "application/x-protobuf"
+)
+
+type worker struct {
+	authToken  string
+	client     *http.Client
+	endpoint   string
+	gzipWriter *gzip.Writer
+	maxRetries uint64
+	doneCh     chan struct{}
+}
+
+func newWorker(client *http.Client, endpoint string, authToken string, maxRetries uint64) *worker {
+	w := &worker{
+		authToken:  authToken,
+		client:     client,
+		endpoint:   endpoint,
+		gzipWriter: gzip.NewWriter(nil),
+		maxRetries: maxRetries,
+		doneCh:     make(chan struct{}),
+	}
+	return w
+}
+
+func (w *worker) export(ctx context.Context, batch *jaegerpb.Batch) error {
+	_, span := trace.StartSpan(ctx, "export")
+	defer span.End()
+
+	if len(batch.Spans) == 0 {
+		span.AddAttributes(trace.BoolAttribute("empty_batch", true))
```

Comment:
Not particularly important. I was trying to imagine a trace in Omnition that dropped data and felt having this attribute to tell that data was dropped because the batch was empty would be useful. On second thought, I think it would be more useful to just have the number of spans always added as an attribute.